Adding the tpcc specific benchmark.

psudheer21 opened the pull request:
Summary:
This change is a fork of oltpbench with all the other benchmarks removed
and using the yugabyte specific driver for communication with our
database.

1. The ivy.xml and pom.xml has the dependency changes.
2. The DBWorkload.java has some changes so that tpcc is the only benchmark
   that is supported.
3. README is modified with the required details to download/compile and
   use the tool.
4. config has the required changes to run the tool against our DB.

Reviewers: Karthik, Mihnea

psudheer21 committed:
Adding the tpcc specific benchmark.

Summary:
This change is a fork of oltpbench with all the other benchmarks removed
and using the yugabyte specific driver for communication with our
database.

1. The ivy.xml and pom.xml has the dependency changes.
2. The DBWorkload.java has some changes so that tpcc is the only benchmark
   that is supported.
3. README is modified with the required details to download/compile and
   use the tool.
4. config has the required changes to run the tool against our DB.

Reviewers: Karthik, Mihnea

ndeodhar commented on README.md:
`````diff
@@ -1,2 +1,82 @@
+````xml
+<!-- config/sample_tpcc_config.xml -->
+    <!-- Connection details -->
+    <dbtype>mysql</dbtype>
`````
Lets make this Yugabyte specific.

psudheer21 replied:
Done.

ndeodhar commented on README.md:
````diff
@@ -1,2 +1,82 @@
+The following command for example initiate a tpcc database (--create=true --load=true) and a then run a workload as described in config/sample_tpcc_config.xml file. The results (latency, throughput) are summarized into 5seconds buckets (-s 5) and the output is written into two file: outputfile.res (aggregated) and outputfile.raw (detailed):
+
+```
+./tpccbenchmark -c config/sample_tpcc_config.xml --create=true --load=true --execute=true -s 5 -o outputfile
````
Same: use YB config file in example.
As a next step, we can make `-c` optional and use a default YB tpcc config file.

psudheer21 replied:
Done.

ndeodhar commented on README.md:
````diff
@@ -1,2 +1,82 @@
+  ```
+
+## Setup of the Database
+If you start from scratch, you should create an empty database (e.g., for TPC-C you can create a db named <b>tpcc</b>) and provide login credential to the benchmark, by modifying  <b>the workload descriptor file</b>. The ./config directory provides several examples, we now use the sample_tpcc_config.xml. You should edit the following portion:
````
Please provide ysql commands that users will need to run for this.

psudheer21 replied:
There are no ysql commands necessary.
IT creates the DB/table and inserts into the table in the load step.

ndeodhar commented on config/yugabyte_tpcc_config_1.xml:
```diff
@@ -0,0 +1,40 @@
+<?xml version="1.0"?>
```
I would recommend making the config file name more descriptive. For example, `config_workload_1`, `config_workload_2` etc. We can also remove the `yugabyte_tpcc` prefix since this YB tpcc repo anyway.

Lets also add one more config to run all workloads (similar to what oltpbench has) - `config_workload_all`

psudheer21 replied:
Done.

ndeodhar reviewed (changes requested):
Great to see this @psudheer21 !
Had some initial comments on this.

psudheer21 committed:
Changes to README and configs

Summary:
Added Yugabyte specific instructions in the README.
Made changes to config to be Yugabyte specific.

Reviewers: Karthik, Mihnea, Neha

psudheer21 committed:
Changes to README and configs

Summary:
Added Yugabyte specific instructions in the README.
Made changes to config to be Yugabyte specific.

Reviewers: Karthik, Mihnea, Neha

ndeodhar commented on README.md:
```diff
@@ -1,2 +1,82 @@
-Repo to run TPCC benchmarks against YugabyteDB
+# TPCC benchmark
+
+This is a fork of Oltp bench that is used to run the TPCC benchmark. All the other benchmarks have been removed.
```
nit: 
```suggestion
This is a fork of OLTPBench that is used to run the TPCC benchmark. All the other benchmarks have been removed.
```

psudheer21 replied:
done.

ndeodhar commented on README.md:
```diff
@@ -1,2 +1,82 @@
+    <DBUrl>jdbc:postgresql://<ip>:5433/yugabyte</DBUrl>
+    <username>yugabyte</username>
+    <password></password>
+    <isolation>TRANSACTION_READ_COMMITTED</isolation>
```
Lets make this REPEATABLE_READ instead of READ_COMMITTED since that's the Yugabyte default. As a side note, we bump up "read committed" to be "repeatable read" anyway.

psudheer21 replied:
done.
Thanks for the explanation. :)

ndeodhar commented on README.md:
`````diff
@@ -1,2 +1,82 @@
+    <isolation>TRANSACTION_READ_COMMITTED</isolation>
+````
+
+The details of the workloads has already been populated in the sample config present in /config.
`````
```suggestion
The details of the workloads have already been populated in the sample config present in /config.
```

psudheer21 replied:
done

ndeodhar commented on README.md:
`````diff
@@ -1,2 +1,82 @@
+````
+
+The details of the workloads has already been populated in the sample config present in /config.
+The workload descriptor works the same way as it is in the upstream branch and details can be found in the [on-line documentation](https://github.com/oltpbenchmark/oltpbench/wiki).
`````
```suggestion
The workload descriptor works the same way as it does in the upstream branch and details can be found in the [on-line documentation](https://github.com/oltpbenchmark/oltpbench/wiki).
```

psudheer21 replied:
done

ndeodhar commented on README.md:
````diff
@@ -1,2 +1,82 @@
+```
+
+## Example
+The following command for example initiate a tpcc database (--create=true --load=true) and a then run a workload as described in config/workload_1.xml file. The results (latency, throughput) are summarized into 5 seconds buckets (-s 5) and the output is written into two file: outputfile.res (aggregated) and outputfile.raw (detailed):
````
```suggestion
The following command, for example, initiates a tpcc database (--create=true --load=true) and then runs a workload as described in config/workload_1.xml file. The results (latency, throughput) are summarized into 5 seconds buckets (-s 5) and the output is written into two file: outputfile.res (aggregated) and outputfile.raw (detailed):
```

psudheer21 replied:
done.

ndeodhar commented on README.md:
````diff
@@ -1,2 +1,82 @@
+./tpccbenchmark -c config/workload_1.xml --create=true --load=true --execute=true -s 5 -o outputfile
+```
+
+Since data loading can be a lengthy process, one would first create a and populate a database which can be reused for multiple experiments:
````
```suggestion
Since data loading can be a lengthy process, one could first create and populate a database which can then be reused for multiple experiments:
```

psudheer21 replied:
done

ndeodhar commented on config/workload_1.xml:
```diff
@@ -0,0 +1,40 @@
+    <DBUrl>jdbc:postgresql://127.0.0.1:5433/yugabyte</DBUrl>
+    <username>yugabyte</username>
+    <password></password>
+    <isolation>TRANSACTION_READ_COMMITTED</isolation>
```
Same comment here (and other config files): Lets use REPEATABLE READ

psudheer21 replied:
done.

ndeodhar commented on config/workload_all.xml:
```diff
@@ -0,0 +1,40 @@
+          <time>30</time>
+          <rate>10000</rate>
+          <ratelimited bench="tpcc">true</ratelimited>
+          <weights>45,43,3,3,3</weights>
```
```suggestion
          <weights>45,43,4,4,4</weights>
```

psudheer21 replied:
done.

ndeodhar reviewed (approved):
There are some files like:
src/com/oltpbenchmark/benchmarks/tpcc/ddls/tpcc-memsql-ddl.sql
src/com/oltpbenchmark/benchmarks/tpcc/dialects/tpcc-memsql-dialects.xml
which we don't need and can remove.

Overall, LGTM. Had a few comments, please fix those before merging.

psudheer21 committed:
Merge branch 'new_tpcc' of github.com:yugabyte/tpcc into new_tpcc

psudheer21 commented:
With 1024 elements and all the changes done as well. I get really similar results:
ysql_bench_batch_sizing:yugabyte-db $ time  build/latest/postgres/bin/ysql_bench -h 127.0.0.1 -p 5433 -i -s 20 --batch-size 1024 yugabyte
dropping old tables...
NOTICE:  table "ysql_bench_accounts" does not exist, skipping
NOTICE:  table "ysql_bench_branches" does not exist, skipping
NOTICE:  table "ysql_bench_history" does not exist, skipping
NOTICE:  table "ysql_bench_tellers" does not exist, skipping
creating tables (with primary keys)...
generating data...
100000 of 2000000 tuples (5%) done (elapsed 5.76 s, remaining 109.45 s)
200000 of 2000000 tuples (10%) done (elapsed 12.46 s, remaining 112.16 s)
...
1800000 of 2000000 tuples (90%) done (elapsed 151.12 s, remaining 16.79 s)
1900000 of 2000000 tuples (95%) done (elapsed 161.92 s, remaining 8.52 s)
2000000 of 2000000 tuples (100%) done (elapsed 173.04 s, remaining 0.00 s)
done.

real	3m0.322s
user	0m1.435s
sys	0m0.290s
